### PR TITLE
Level down the lock of creating indexes on AO tables if blkdir exists

### PR DESCRIPTION
**Pull request by adam8157**
Appendoptimized tables need block directory relation for index access. Creating
and maintaining a block directory is expensive. We delay the block directory
creation until it is really needed - the first index creation. Once created,
all indexes share the same block directory. We need a stronger lock
(ShareRowExclusiveLock) if the block directory needs to be created. If the
block directory already exists, we can use the same lock as heap tables.

### CODE REVIEW

**Comment by kainwen** [comment]
Can this fix https://github.com/greenplum-db/gpdb/issues/10961 ?

Also, does this pr still contain lock level upgrade during create index (the first time)?

**Comment by adam8157** [comment]
> Can this fix #10961 ?
> 
> Also, does this pr still contain lock level upgrade during create index (the first time)?

Partially, the first time still needs `ShareRowExclusiveLock`.

**Comment by ashwinstar** [comment]
PR pipeline failure is pointing with the change, for partition tables we are taking more locks.

File: src/backend/commands/indexcmds.c [hunk 1]
```diff
@@ -680,26 +681,32 @@ DefineIndex(Oid relationId,
...
+		 * (ShareRowExclusiveLock) that blocks index creation from another
+		 * transaction (not to be confused with create index concurrently) as
+		 * well as concurrent insert for appendoptimized tables, if the block
+		 * directory needs to be created. If the block directory already exists,
+		 * we can use the same lock as heap tables.
+		 */
+		if (!OidIsValid(blkdirrelid))
+			lockmode = ShareRowExclusiveLock; // Relation is AO, and has no block directory
 		else
 			lockmode = stmt->concurrent ? ShareUpdateExclusiveLock : ShareLock;
```
Comment (ashwinstar) [inline src/backend/commands/indexcmds.c:708]: How is `else` case triggered for HEAP or any other table type (blkdirrelid = InvalidOid) for it as well? Maybe we should write

```
lockmode = stmt->concurrent ? ShareUpdateExclusiveLock : ShareLock;
if (RelationIsAppendOptimized(rel))
{
    Oid blkdirrelid = InvalidOid;
    GetAppendOnlyEntryAuxOids(relationId, NULL, NULL, &blkdirrelid, NULL, NULL, NULL);
    if (!OidIsValid(blkdirrelid))
       lockmode = ShareRowExclusiveLock;
}

Comment (adam8157) [reply src/backend/commands/indexcmds.c]: Thanks, fixed.

File: src/backend/commands/indexcmds.c [hunk 2]
```diff
@@ -680,28 +681,32 @@ DefineIndex(Oid relationId,
...
-	 * the same lock as heap tables but this is not currently implemented
-	 * (hence the fixme).
+	 * transaction (not to be confused with create index concurrently) as
+	 * well as concurrent insert for appendoptimized tables, if the block
+	 * directory needs to be created. If the block directory already exists,
+	 * we can use the same lock as heap tables.
 	 */
+	rel = table_open(relationId, NoLock);
+	if (RelationIsAppendOptimized(rel))
 	{
```
Comment (asimrp) [inline src/backend/commands/indexcmds.c:701]: This anonymous scope was useful to delimit Greenplum-specific changes, my personal preference is to keep it.  Of course, feel free to disregard the comment.

**Review by asimrp — APPROVED** [review]

**Pull request merged**